### PR TITLE
RATIS-1619. Validate server id in raft group if the group is not empty

### PR DESCRIPTION
**Pull request by tisonkun**
## What changes were proposed in this pull request?

Validate server id in raft group if the group is not null

## What is the link to the Apache JIRA

[RATIS-1619](https://issues.apache.org/jira/browse/RATIS-1619)

## How was this patch tested?

Add new tests.

### CODE REVIEW

**Comment by tisonkun** [comment]
cc @szetszwo @codings-dan

**Comment by tisonkun** [comment]
It seems `org.apache.ratis.server.impl.MiniRaftCluster#putNewServers` depends on the group that doesn't contains the new peer. I have no idea yet whether it's a legacy logic we should update here or the original purpose of this ticket is invalid.

**Comment by tisonkun** [comment]
`org.apache.ratis.server.impl.RaftReconfigurationBaseTest#testReconfTwice` can be unstable with current version. Investigating...

**Comment by tisonkun** [comment]
Tests failed at:

```
Error:    TestServerRestartWithSimulatedRpc>ServerRestartTests.testRestartWithCorruptedLogEntryWithException:345->BaseTest.testFailureCase:167->BaseTest.testFailureCase:159 The test "restart-fail-ChecksumException" does not throw anything.
Error:  Errors: 
Error:    TestNettyDataStreamChainTopologyWithGrpcCluster.testMultipleStreamsMultipleServersStepDownLeader » Completion
```

* `TestServerRestartWithSimulatedRpc>ServerRestartTests.testRestartWithCorruptedLogEntryWithException` cannot be reproduced locally.
* `TestNettyDataStreamChainTopologyWithGrpcCluster.testMultipleStreamsMultipleServersStepDownLeader` failed on master also.

Thus I guess they are flaky tests. Please confirm.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java [hunk 1]
```diff
@@ -47,14 +50,12 @@ public static RaftServerProxy newRaftServer(
...
-    final RaftServerProxy proxy = newRaftServer(id, stateMachineRegistry, properties, parameters);
-    proxy.initGroups(group);
-    return proxy;
-  }
 
-  private static RaftServerProxy newRaftServer(
-      RaftPeerId id, StateMachine.Registry stateMachineRegistry, RaftProperties properties, Parameters parameters)
-      throws IOException {
+    if (group != null && !group.getPeers().isEmpty()) {
+      final Set<RaftPeerId> ids = group.getPeers().stream().map(RaftPeer::getId).collect(Collectors.toSet());
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java:55]: Use RaftGroup.getPeer(id) instead of creating a set.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java [hunk 2]
```diff
@@ -47,14 +50,12 @@ public static RaftServerProxy newRaftServer(
...
       RaftProperties properties, Parameters parameters) throws IOException {
     RaftServer.LOG.debug("newRaftServer: {}, {}", id, group);
-    final RaftServerProxy proxy = newRaftServer(id, stateMachineRegistry, properties, parameters);
-    proxy.initGroups(group);
-    return proxy;
-  }
 
-  private static RaftServerProxy newRaftServer(
-      RaftPeerId id, StateMachine.Registry stateMachineRegistry, RaftProperties properties, Parameters parameters)
-      throws IOException {
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java:57]: Please revert the unrelated changes.  The original code is more clear.

**Review by szetszwo — CHANGES_REQUESTED** [review]
@tisonkun , thanks for working on this!  Some comments inlined.  The failure of TestPeerCommandIntegrationWithGrpc is related to this.  Please take a look.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java (hunk 1 above)
Comment (tisonkun) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java:55]: Thanks! I don't notice this method :P

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java (hunk 2 above)
Comment (tisonkun) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ServerImplUtils.java:57]: Reverted.

**Comment by tisonkun** [comment]
> The failure of TestPeerCommandIntegrationWithGrpc is related to this.

I'm pushing a commit to address style comments. Will investigate this case now.

**Comment by tisonkun** [comment]
```
Error:  Failures: 
Error:    TestLeaderElectionWithSimulatedRpc>LeaderElectionTests.testEnforceLeader:235->LeaderElectionTests.enforceLeader:249 expected:<s[1]> but was:<s[3]>
Error:    TestServerRestartWithSimulatedRpc>ServerRestartTests.testRestartWithCorruptedLogEntryWithException:345->BaseTest.testFailureCase:167->BaseTest.testFailureCase:159 The test "restart-fail-ChecksumException" does not throw anything.
Error:  Errors: 
Error:    TestNettyDataStreamChainTopologyWithGrpcCluster>DataStreamAsyncClusterTests.testMultipleStreamsMultipleServersStepDownLeader:78->DataStreamAsyncClusterTests.runTestDataStreamStepDownLeader:82->DataStreamAsyncClusterTests.runTestDataStream:97 » TestTimedOut
Error:    TestNettyDataStreamStarTopologyWithGrpcCluster.testMultipleStreamsMultipleServers » Completion
Error:    TestNettyDataStreamStarTopologyWithGrpcCluster>DataStreamAsyncClusterTests.lambda$runTestDataStream$2:106 » Completion
```

* `TestLeaderElectionWithSimulatedRpc>LeaderElectionTests.testEnforceLeader` cannot reproduce locally.
* `TestServerRestartWithSimulatedRpc>ServerRestartTests.testRestartWithCorruptedLogEntryWithException` cannot reproduce locally, reported above.
* `TestNettyDataStreamChainTopologyWithGrpcCluster.testMultipleStreamsMultipleServersStepDownLeader` failed on master also, reported above.

File: ratis-common/src/main/java/org/apache/ratis/util/Preconditions.java [hunk 3]
```diff
@@ -102,6 +102,11 @@ static <T> T assertNotNull(T object, String name) {
         + name + " = " + object + " == null, class = " + object.getClass());
   }
 
+  static <T> T assertNotNull(T object, String format, Object... args) {
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/util/Preconditions.java:105]: FYI, we usually use  `Objects.requireNonNull(..)`.  It is fine if we add this method.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

File: ratis-common/src/main/java/org/apache/ratis/util/Preconditions.java (hunk 3 above)
Comment (tisonkun) [reply ratis-common/src/main/java/org/apache/ratis/util/Preconditions.java:105]: Good to know. I made the decision since `Objects.requireNonNull` throws `NullPointerException` while these methods throws `IllegalStateException`. Perhaps Guava's `Precondition.checkArgument` fits better XD
